Docs: Add CSF factory API documentation (#30343)
storybookjs/storybook: docs/api/csf/csf-factories.mdx

---
title: 'Component Story Format (CSF)'
isTab: true
tab:
  order: 2
  title: CSF Factories (Experimental)
---

<If notRenderer="react">

<Callout variant="info">
  CSF Factories are currently only supported in [React](?renderer=react) projects.
</Callout>

</If>

<If renderer="react">

<Callout variant="warning" icon="🧪">
  This is an experimental feature and (though unlikely) the API may change in future releases. We welcome feedback and contributions to help improve this feature.
</Callout>

CSF Factories are the next evolution of Storybook's Component Story Format (CSF). This new API uses a pattern called factory functions to provide full type safety to your Storybook stories, making it easier to configure addons correctly and unlocking the full potential of Storybook's features.

This reference will provide an overview of the API and a migration guide to upgrade from CSF 3.

## Overview

> The CSF Factories API is composed of four main functions to help you write stories. Note how the functions operate as a factory, each producing the next function in the chain (`definePreview` → `preview.meta` → `meta.story`), providing full type safety at each step.

Author: ```suggestion
The CSF Factories API is composed of three main functions to help you write stories. Note how the functions operate as factories, each producing the next function in the chain (`definePreview` → `preview.meta` → `meta.story`), providing full type safety at each step.
```

### `defineMain`

With CSF Factories, your [main Storybook config](../main-config/main-config.mdx) is specified by the `defineMain` function. This function is type-safe and will automatically infer types for your project.

{/* prettier-ignore-start */}
<CodeSnippets path="csf-factories-main-example.md" />
{/* prettier-ignore-end */}

> ### `definePreview`

Author: would it make sense to mention in this section that now the preview object will be used to define stories, and therefore will be imported in every story file?

{/* TODO: Snippetize */}
```ts title="Button.stories.ts"
// Learn about the # subpath import: https://storybook.js.org/docs/api/csf/csf-factories#subpath-imports
import preview from '#.storybook/preview';

const meta = preview.meta({
  component: Button,
  // ...
});

export const Primary = meta.story({
  args: {
    // type-safe!
    primary: true,
  },
});
```

### Subpath imports

CSF Factories leverages subpath imports to simplify importing constructs from the preview file. While you can still use relative path imports, subpath imports offer a more convenient and maintainable approach:

```ts
// ✅ Subpath imports won't break if you move story files around
import preview from '#.storybook/preview';

// ❌ Relative imports will break if you move story files around
import preview from '../../../.storybook/preview';
```

The [CSF Factories codemod](#codemod) will automatically add the necessary subpath import for the preview file, or you can [add it manually](#1-add-subpath-import-in-packagejson).

For more details, refer to the [subpath imports documentation](../../writing-stories/mocking-data-and-modules/mocking-modules.mdx#subpath-imports).

## Upgrading from CSF 3

> You can upgrade your project's story files to CSF Factories incrementally or all at once. However, you must upgrade your `.storybook/main.ts` and `.storybook/preview.ts` files first before you can use CSF Factories in a story file.

Author: FYI the main file doesn't really have anything to do with csf factory, it's just a type helper that follows the `define` pattern. Though not needed, it's a good thing to have users migrate to the new format anyway!

Reviewer: Yeah, I know. It's just more clear to do the less nuanced thing here.

<Callout variant="info">
  If your story files are not using CSF 3 already, please run `npx storybook migrate csf-2-to-3` to automatically upgrade your project from CSF 2 to CSF 3.

  You can also [migrate manually](./index.mdx#upgrading-from-csf-2-to-csf-3).

> </Callout>

Author: We can now change/remove this section! The codemod will also migrate from CSF 2 to 3.

Reviewer: And CSF 1, presumably? Because that's just CSF 2 without `args`?

### Codemod

Storybook provides codemods to help you migrate your codebase to the CSF Factories format effortlessly. Run the following command in the root of your project: 

```sh
npx storybook automigrate csf-factories
```

<Callout variant="info" icon="💡">
  If you use a monorepo or your `.storybook` directory is not at the root, make sure to pass the `--config-dir` flag to the command, specifying its path.
</Callout>

This command performs the following actions:

- Adds [a subpath import](#1-add-subpath-import-in-packagejson) for the preview file in your `package.json`
- [Migrates `.storybook/main.js`](#2-update-your-main-storybook-config-file)
- [Migrates `.storybook/preview.js`](#3-update-your-preview-config-file)
- Converts all story files to [use CSF Factories](#4-update-your-story-files)

> This should get you started with CSF Factories immediately. If you prefer a manual migration, the next sections detail the neccessary updates.

Author: ```suggestion
This should get you started with CSF Factories immediately. If you prefer a manual migration, the next sections detail the necessary updates.
```

### 1. Add subpath import in `package.json`

To be able to consistently import the preview file from any location in your project, you need to add a subpath import in your `package.json`. For more information, refer to the [subpath imports documentation](../../writing-stories/mocking-data-and-modules/mocking-modules.mdx#subpath-imports).

{/* TODO: Snippetize */}
```json title="package.json"
{
  "imports": {
    "#.storybook/preview": {
      "default": "./.storybook/preview.ts",
    },

>     "#*": ["./*", "./*.ts", "./*.tsx"],

Author: Interesting 🤔 the codemod only adds `"#*": ["./*", "./*.ts", "./*.tsx"]` and assumes that users will write `#.storybook/preview` which will relate to `<root>.storybook/preview` and if they have the preview file somewhere else, they will have to adjust the import in each file to something like `#some-dir/.storybook/preview`.

This means users with multiple Storybooks could have multiple imports like `#apps/.storybook/preview` and `#libs/.storybook/preview` without having to change anything in their package.json.

Perhaps the codemod should set this more specific field like what Kyle wrote here instead? WDYT @kasperpeulen?

Reviewer: I see! Thanks for explaining. That's tricky. I see why you did it the current way, but think the way I documented is perhaps a bit more informative? I like that if someone does a Find for `#.storybook/preview`, they'll come across this.

Author: It's tricky, I'll leave @kasperpeulen to decide

Reviewer: @kasperpeulen — Did you have thoughts? I'm still inclined to think it should be the way I've documented here, because:

> I like that if someone does a Find for `#.storybook/preview`, they'll come across this.

Reviewer 2: I'm not sure, I will let @shilman decide about this. :P

Reviewer: Fwiw, I'm 60/40 on my proposal here vs. what was implemented. So unless someone else strongly agrees with me, we can just keep it simple.

Reviewer 3: I can see both sides. I prefer @kylegach 's proposal slightly in the name of being explicit. If we want to support multiple preview files like you describe @yannbf then we should make that explicit instead and document it here.

  },
}
```

### 2. Update your main Storybook config file

Update your `.storybook/main.ts` file to use the new [`defineMain`](#definemain) function.

```diff title=".storybook/main.ts"
// Replace your-framework with the framework you are using (e.g., react-vite, nextjs, experimental-nextjs-vite)
+ import { defineMain } from '@storybook/your-framework/node';
- import { StorybookConfig } from '@storybook/your-framework';

+ export default defineMain({
- export const config: StorybookConfig = {
    // ...current config
+ });
- };
- export default config;
```

### 3. Update your preview config file

Update your `.storybook/preview.ts` file to use the new [`definePreview`](#definepreview) function.

```diff title=".storybook/preview.ts"
// Replace your-framework with the framework you are using (e.g., react-vite, nextjs, experimental-nextjs-vite)

> + import { definePreview } from '@storybook/your-framework/node';

Author: ```suggestion
+ import { definePreview } from '@storybook/your-framework';
```

- import { Preview } from '@storybook/your-renderer';
// 👇 Import the addons you are using

> + import addonTest from '@storybook/experimental-addon-test';

Author: This part is quite tricky. Our core addons will be a function, a default export and part of the root entrypoint. Older addons will not have that format, the addon-kit also doesn't have that format (yet). The codemod will figure everything out to fill things properly, with this logic:
- Is it a core addon? Does it contain (./preview) annotations? Then it's default imported as a function
- If not a core addon, does it contain annotations? Then it's namespace imported as object

So in this scenario:
```ts
// .storybook/main

export default {
  addons: [
     '@storybook/addon-onboarding', // core addon, without annotations - then it's ignored
     '@storybook/experimental-addon-test', // core addon, with annotations
     'storybook-addon-designs', // non-core addon, without annotations - then it's ignored
     'storybook-addon-performance', // non-core addon, with annotations
  ]
}
```

the preview file would become:
```ts
// .storybook/preview

import { definePreview } from '@storybook/your-framework';
import experimentalAddonTest from '@storybook/experimental-addon-test';
import * as storybookAddonPerformance from 'storybook-addon-performance/preview';

export default definePreview({
  addons: [
    experimentalAddonTest(),
    storybookAddonPerformance,
  ],
});
```

Author: In the future hopefully all addons will use the same format with functions, but it will be difficult to enforce that. There will be a long transition period from the older format to the new one

Reviewer: That's a _super_ helpful explanation; thank you! I'll spend some time on this to clarify more cases.

Reviewer: @yannbf — I took a stab at this. Mind reviewing, please?

- type Story = StoryObj<typeof meta>;

+ export const Primary = meta.story({
- export const Primary: Story = {
    // ...current story
+ });
- };
```

Note how it is no longer necessary to import any types or manually apply types to the meta or stories. The types are now inferred automatically, thanks to the factory function pattern.

#### 4.1 Reusing story properties

Previously, story properties such as `Story.args` or `Story.parameters` were accessed directly when reusing them in another story. While accessing them like this is still supported, it is deprecated in CSF Factories.

All of the story properties are now contained within a new property called `composed` and should be accessed from that property instead. For instance, `Story.composed.args` or `Story.composed.parameters`.

```diff title="Button.stories.ts"
// ...rest of file

+ export const Primary = meta.story({
- export const Primary: Story = {
    args: { primary: true },
+ });
- };

+ export const PrimaryDisabled = meta.story({
- export const PrimaryDisabled: Story = {
    args: {
+     ...Primary.composed.args,
-     ...Primary.args,
      disabled: true,
    }
+ });
- };
```

<Callout variant="info">

>   The property name "composed" was chosen because the values within are composed from the story, its component meta, and the preview configuration.

Author: btw the codemod will also change story files like this and will use `.input` instead of `.composed`. I'm assuming that users who originally wrote `MyStory.args` are expecting to see the args that they wrote in the story, not the composed ones coming from preview/meta as well. Though I'd say it's more valuable to get the composed ones!

import { screen } from '@testing-library/react';
- import { composeStories } from '@storybook/react';
 
// Import all stories from the stories file
import * as stories from './Button.stories';
 
+ const { Primary } = stories;
- const { Primary } = composeStories(stories);
 
test('renders primary button with default args', async () => {
  // The run function will mount the component and run all of Storybook's lifecycle hooks
  await Primary.run();
  const buttonElement = screen.getByText('Text coming from args in stories file!');
  expect(buttonElement).not.toBeNull();
});
```

The `Story` object also provides a `Component` property in case you want to render it using another method of your choosing, for instance, with [Testing Library](https://testing-library.com/). You also have access to its composed properties (args, parameters, etc.) via the `composed` property.

Here's an example of how you can reuse a story in a test file by rendering its component:

{/* prettier-ignore-start */}
<CodeSnippets path="portable-stories-csf-factory-render.md" />
{/* prettier-ignore-end */}

## Frequently asked questions (FAQ)

### Will I have to migrate all of my stories to this new format?

> Storybook will continue to support CSF 1, [CSF 2](../../../release-6-5/docs/api/stories/csf.mdx), and [CSF 3](./index.mdx) for the foreseeable future. None of these prior formats are deprecated.

Author: ```suggestion
Storybook will continue to support CSF 1, [CSF 2](../../../release-6-5/docs/api/stories/csf.mdx), and [CSF 3](./index.mdx) for the foreseeable future. None of these prior formats is deprecated.
```

Reviewer: This is incorrect. The deprecated state applies to the formats, which is plural, so "are" is correct.

While using CSF Factories, you can still use the older formats, as long as they are not mixed in the same file. If you want to migrate your existing files to the new format, refer to [the upgrade section](#upgrading-from-csf-3), above.

> ### How can I know more about this format and provide feedback?

Author: Perhaps we add one more:

```
### Will this format work with docs(.mdx) pages?

Yes, the docs blocks support the CSF factory format and you wouldn't have to change the code in your mdx files.
```

</Callout>

CSF Factories are the next evolution of Storybook's Component Story Format (CSF). This new API uses a pattern called factory functions to provide full type safety to your Storybook stories, making it easier to configure addons correctly and unlocking the full potential of Storybook's features.

This reference will provide an overview of the API and a migration guide to upgrade from CSF 3.

## Overview

The CSF Factories API is composed of four main functions to help you write stories. Note how three of the functions operate as factories, each producing the next function in the chain (`definePreview` → `preview.meta` → `meta.story`), providing full type safety at each step.

### `defineMain`

With CSF Factories, your [main Storybook config](../main-config/main-config.mdx) is specified by the `defineMain` function. This function is type-safe and will automatically infer types for your project.

{/* prettier-ignore-start */}
<CodeSnippets path="csf-factories-main-example.md" />
{/* prettier-ignore-end */}

### `definePreview`

> Similarly, your project's stories configuration is specified by the `definePreview` function. This is also type-safe and will infer types throughout your project.

Reviewer 4: Words not precious, but we can go with something as follows for clarity
```suggestion
Similarly, the `definePreview` function specifies your project's story configuration. This function is also type-safe and will infer types throughout your project.
```

{/* TODO: Snippetize */}
```ts title="Button.stories.ts"
// Learn about the # subpath import: https://storybook.js.org/docs/api/csf/csf-factories#subpath-imports
import preview from '#.storybook/preview';

const meta = preview.meta({
  component: Button,
  // ...
});

export const Primary = meta.story({
  args: {
    // type-safe!
    primary: true,
  },
});
```

### Subpath imports

CSF Factories leverages subpath imports to simplify importing constructs from the preview file. While you can still use relative path imports, subpath imports offer a more convenient and maintainable approach:

```ts
// ✅ Subpath imports won't break if you move story files around
import preview from '#.storybook/preview';

// ❌ Relative imports will break if you move story files around
import preview from '../../../.storybook/preview';
```

The [CSF Factories codemod](#codemod) will automatically add the necessary subpath import for the preview file, or you can [add it manually](#1-add-subpath-import-in-packagejson).

For more details, refer to the [subpath imports documentation](../../writing-stories/mocking-data-and-modules/mocking-modules.mdx#subpath-imports).

## Upgrading from CSF 1, 2, or 3

> You can upgrade your project's story files to CSF Factories incrementally or all at once. However, you must upgrade your `.storybook/main.ts` and `.storybook/preview.ts` files first before you can use CSF Factories in a story file.

Reviewer 4: Minor rewording here when you can
```suggestion
You can upgrade your project's story files to CSF Factories incrementally or all at once. However, before using CSF Factories in a story file, you must upgrade your `.storybook/main.ts` and `.storybook/preview.ts` files.
```

+ import { defineMain } from '@storybook/your-framework/node';
- import { StorybookConfig } from '@storybook/your-framework';

+ export default defineMain({
- export const config: StorybookConfig = {
    // ...current config
+ });
- };
- export default config;
```

### 3. Update your preview config file

Update your `.storybook/preview.ts` file to use the new [`definePreview`](#definepreview) function.

<details>
<summary>Which addons should be specified in `preview`?</summary>

The ability for an addon to provide annotation types (`parameters`, `globals`, etc.) is new and not all addons support it yet. 

If an addon provides annotations (i.e. it distributes a `./preview` export), it can be imported in two ways:

1. For official Storybook addons, you import the default export:
   `import addonName from '@storybook/addon-name'`

2. For community addons, you should import the entire module and access the addon from there:
   `import * as addonName from 'community-addon-name'`

Sound complicated? The [codemod](#codemod) will automatically add the necessary imports for you.   

</details>

> ```diff title=".storybook/preview.ts"

Reviewer 4: minor item here for consistency, add js as well, turning this into `storybook/preview.js|ts`) and apply it to other places as well

- };
```

Note how it is no longer necessary to import any types or manually apply types to the meta or stories. The types are now inferred automatically, thanks to the factory function pattern.

#### 4.1 Reusing story properties

Previously, story properties such as `Story.args` or `Story.parameters` were accessed directly when reusing them in another story. While accessing them like this is still supported, it is deprecated in CSF Factories.

All of the story properties are now contained within a new property called `composed` and should be accessed from that property instead. For instance, `Story.composed.args` or `Story.composed.parameters`.

```diff title="Button.stories.ts"
// ...rest of file

+ export const Primary = meta.story({
- export const Primary: Story = {
    args: { primary: true },
+ });
- };

+ export const PrimaryDisabled = meta.story({
- export const PrimaryDisabled: Story = {
    args: {
+     ...Primary.composed.args,
-     ...Primary.args,
      disabled: true,
    }
+ });
- };
```

<Callout variant="info">
  The property name "composed" was chosen because the values within are composed from the story, its component meta, and the preview configuration.

>   If you wish instead to access only the direct input to the story, you can use `Story.input` instead of `Story.composed`.

Reviewer 4: We can go with something like this for clarity
```suggestion
 If you want to access the direct input to the story, you can use `Story.input` instead of `Story.composed`.
```

  When running the [codemod](#codemod), it will automatically update usages of `Story.args` to `Story.input.args`. This is to ensure that your stories continue to work as expected. However, using `Story.composed.args` is likely more helpful in most cases.
</Callout>

### 5. Reusing stories in test files

> [Storybook's Test addon](../../writing-tests/test-addon.mdx) allows you to test your components directly inside Storybook. All of the stories are automatically turned into Vitest tests, so the integration is seamless in your testing suite.

Reviewer 4: Minor change here for clarity
```suggestion
[Storybook's Test addon](../../writing-tests/test-addon.mdx) allows you to test your components directly inside Storybook. All the stories are automatically turned into Vitest tests, making integration seamless in your testing suite.
```

If you cannot use Storybook Test, you can still reuse the stories in your test files using [portable stories](../portable-stories/portable-stories-vitest.mdx). In prior story formats, you had to compose the stories before rendering them in your test files. With CSF Factories, you can now reuse the stories directly.

```diff title="Button.test.ts"
import { test, expect } from 'vitest';
import { screen } from '@testing-library/react';
- import { composeStories } from '@storybook/react';
 
// Import all stories from the stories file
import * as stories from './Button.stories';
 
+ const { Primary } = stories;
- const { Primary } = composeStories(stories);
 
test('renders primary button with default args', async () => {
  // The run function will mount the component and run all of Storybook's lifecycle hooks
  await Primary.run();
  const buttonElement = screen.getByText('Text coming from args in stories file!');
  expect(buttonElement).not.toBeNull();
});
```

> The `Story` object also provides a `Component` property in case you want to render it using another method of your choosing, for instance, with [Testing Library](https://testing-library.com/). You also have access to its composed properties (args, parameters, etc.) via the `composed` property.

Reviewer 4: Words are not precious but we could go with something like this for clarity:
```suggestion
The `Story` object also provides a `Component` property, enabling you to render the component with any method you choose, such as [Testing Library](https://testing-library.com/). You can also access its composed properties (`args`, `parameters`, etc.) via the `composed` property.
```

Also, again a reminder to provide some links to the documentation when the rest of the snippets are added

Here's an example of how you can reuse a story in a test file by rendering its component:

{/* prettier-ignore-start */}
<CodeSnippets path="portable-stories-csf-factory-render.md" />
{/* prettier-ignore-end */}

## Frequently asked questions (FAQ)

### Will I have to migrate all of my stories to this new format?

Storybook will continue to support CSF 1, [CSF 2](../../../release-6-5/docs/api/stories/csf.mdx), and [CSF 3](./index.mdx) for the foreseeable future. None of these prior formats are deprecated.

While using CSF Factories, you can still use the older formats, as long as they are not mixed in the same file. If you want to migrate your existing files to the new format, refer to [the upgrade section](#upgrading-from-csf-3), above.

### Will this format work with MDX docs pages?

Yes, the [doc blocks](../../writing-docs/doc-blocks.mdx) used to reference stories in MDX files support the CSF Factories format with no changes needed.

### How can I know more about this format and provide feedback?

> For more information on the original proposal of this experimental format, refer to its [RFC on GitHub](https://github.com/storybookjs/storybook/discussions/30112). We welcome your comments!

Reviewer 4: ```suggestion
For more information on this experimental format's original proposal, refer to its [RFC on GitHub](https://github.com/storybookjs/storybook/discussions/30112). We welcome your comments!
```

---
title: 'Component Story Format (CSF)'
isTab: true
tab:
  order: 2
  title: CSF Factories (Experimental)
---

<If notRenderer="react">

<Callout variant="info">
  CSF Factories are currently only supported in [React](?renderer=react) projects.
</Callout>

</If>

<If renderer="react">

<Callout variant="warning" icon="🧪">

>   This is an experimental feature and (though unlikely) the API may change in future releases. We welcome feedback and contributions to help improve this feature.

Reviewer 3: Link here? https://github.com/storybookjs/storybook/discussions/30112
